### PR TITLE
fix: subscription tree pruning

### LINKED ISSUE #2166: Implement subscription tree branch pruning
## Problem

The subscription tree for contract updates doesn't properly prune branch nodes when their downstream subscribers disconnect.

### Current Behavior

When a peer receives an `Unsubscribed` message, it:
1. Removes the peer from its `subscribers` list
2. Does nothing else

### Expected Behavior

When a downstream subscriber unsubscribes, a branch peer should:
1. Remove the peer from its downstream subscribers
2. Check if it has any remaining downstream subscribers
3. If no downstream subscribers AND not seeding locally (`is_seeding_contract`), send `Unsubscribed` to its upstream peer
4. This cascades up the tree, pruning unnecessary branches

### Root Cause

The current `SeedingManager::subscribers` stores a flat `Vec<PeerKeyLocation>` (soon to be `Vec<SocketAddr>`) that doesn't distinguish between:
- **Upstream**: The peer we subscribed through (toward the root/contract location)
- **Downstream**: Peers that subscribed through us (leaves or other branches)

### Suggested Data Model

\`\`\`rust
struct ContractSubscription {
    upstream: Option<SocketAddr>,       // who we're subscribed to (toward root)
    downstream: Vec<SocketAddr>,        // who subscribed through us
    is_local: bool,                     // local user subscribed (leaf node)
}
\`\`\`

### Impact

Without branch pruning:
- Intermediate nodes stay subscribed longer than needed
- Updates are forwarded to branches with no leaves
- Network resources wasted on unnecessary update propagation

### Context

- Discovered during #2164 Phase 4 planning (connection-based routing refactor)
- The routing refactor changes subscribers to `Vec<SocketAddr>` but keeps the flat structure
- This issue can be addressed independently after the routing refactor lands

### Related

- Part of #2164 peer identity simplification effort
- Subscription tree design: https://freenet.org/news/summary-delta-sync/

[AI-assisted - Claude]

### PR DESCRIPTION
**Pull request by netsirius**
## Problem

The subscription tree for contract updates doesn't properly prune branch nodes when their downstream subscribers disconnect. This causes:

- **Intermediate nodes staying subscribed longer than needed** - When all downstream peers disconnect, the branch peer remains subscribed to its upstream, receiving updates it has no one to forward to
- **Wasted network resources** - Updates are forwarded to branches with no leaves, consuming bandwidth and processing time
- **Memory leaks** - Subscription entries accumulate without cleanup

### Root Cause

The existing `SeedingManager::subscribers` stored a flat `Vec<PeerKeyLocation>` that doesn't distinguish between:
- **Upstream**: The peer we subscribed through (toward the root/contract location)
- **Downstream**: Peers that subscribed through us (leaves or other branches)

Without this distinction, we can't know when to cascade an `Unsubscribed` message up the tree.

## Solution

This PR implements proper subscription tree pruning by:

### 1. Adding Role-Aware Subscription Tracking

Replaced the flat subscriber list with a structured model:

```rust
pub enum SubscriptionRole {
    Upstream,   // The peer we get updates FROM
    Downstream, // Peers that get updates FROM us
}

pub struct SubscriptionEntry {
    peer: PeerKeyLocation,
    role: SubscriptionRole,
}
```

### 2. Tracking Client (WebSocket) Subscriptions

Added `client_subscriptions: DashMap<ContractKey, HashSet<ClientId>>` to track which local clients are subscribed. This is critical because:
- A node should NOT prune if it has local clients subscribed (even with zero downstream peers)
- Only when both downstream peers AND local clients are gone should we notify upstream

### 3. Implementing Cascading Prune Logic

When a downstream peer disconnects or sends `Unsubscribed`:
1. Remove them from downstream subscribers
2. Check if any downstream subscribers remain
3. Check if any local client subscriptions exist
4. If BOTH are empty, return the upstream peer for notification
5. The caller sends `Unsubscribed` to upstream, cascading the pruning up the tree

### 4. Updated API

- `add_downstream(contract, peer, observed_addr)` - Register downstream subscriber
- `set_upstream(contract, peer)` - Set upstream source
- `remove_subscriber(contract, peer) -> RemoveSubscriberResult` - Remove with pruning check
- `prune_subscriptions_for_peer(location) -> PruneSubscriptionsResult` - Bulk cleanup on disconnect
- `add_client_subscription(contract, client_id)` / `remove_client_subscription(...)` - Client tracking

### Why track observed_addr separately?

NAT peers may embed incorrect (loopback) addresses in their `PeerKeyLocation`. The transport-level observed address is the only reliable way to route back to them. The `add_downstream` method accepts an optional `ObservedAddr` that overrides the embedded address.

### Why use DashMap for subscriptions?

Concurrent access from multiple async tasks (subscription handling, connection pruning, client disconnects) requires thread-safe data structures. DashMap provides fine-grained locking without holding locks across await points.

### Why separate client subscription tracking?

Client subscriptions (WebSocket) are fundamentally different from peer subscriptions:
- They don't have a `PeerKeyLocation` (they're local)
- Multiple clients can subscribe to the same contract
- They must be counted to prevent premature pruning

## Testing

Added 44 comprehensive unit tests covering:
- Basic downstream/upstream operations
- Duplicate handling and max subscriber limits
- Client subscription tracking
- Pruning scenarios (with/without downstream, with/without clients)
- Bulk pruning on peer disconnect
- Edge cases (remove non-existent, empty contracts, etc.)

All existing tests continue to pass (494 tests total).

## Files Changed

| File | Changes |
|------|---------|
| `ring/seeding.rs` | Core subscription tree implementation with role tracking |
| `ring/mod.rs` | Updated Ring API to expose new subscription methods |
| `operations/subscribe.rs` | Use `add_downstream`/`set_upstream` API |
| `operations/update.rs` | Use `add_downstream` for broadcast subscribers |
| `operations/put.rs` | Added `PeerKeyLocation` import |
| `client_events/websocket.rs` | Client subscription/unsubscription handling |
| `node/network_bridge/p2p_protoc.rs` | Handle `PruneSubscriptionsResult` on disconnect |
| `test_utils.rs` | Added `make_node_diagnostics` helper |
| `tests/operations.rs` | New integration tests for subscription tree |

## Fixes

Closes #2166

### CODE REVIEW

**Comment by netsirius** [comment]
@claude review it

**Comment by netsirius** [comment]
@claude review this PR again and check if it solves https://github.com/freenet/freenet-core/issues/2166

**Comment by sanity** [comment]
@claude How does this determine which peers are upstream/downstream? is it based on the peer's location relative to this peer's location and the contract location? (ie. if the connected peer is closer to the contract than the current peer then it is upstream)?

**Comment by netsirius** [comment]
@sanity Doesn't the message flow already roughly satisfy this though? The routing uses `k_closest_potentially_caching` to pick the next hop, so messages naturally move toward peers closer to the contract. The flow-based approach just "freezes" that routing decision at subscription time, capturing the path that actually worked (with skip lists, available connections, ...)

File: crates/core/src/ring/mod.rs [hunk 1]
```diff
@@ -187,40 +188,38 @@ impl Ring {
...
     /// The `size_bytes` should be the size of the contract state.
-    pub fn seed_contract(&self, key: ContractKey, size_bytes: u64) -> Vec<ContractKey> {
+    ///
+    /// Returns a list of (evicted_contract, upstream_to_notify) pairs.
+    /// The upstream should be sent an Unsubscribed message for each evicted contract.
+    pub fn seed_contract(
+        &self,
+        key: ContractKey,
+        size_bytes: u64,
+    ) -> Vec<(ContractKey, Option<PeerKeyLocation>)> {
```
Comment (iduartgomez) [inline crates/core/src/ring/mod.rs]: maybe type this into a struct for clarity, instead of returning a tuple

File: crates/core/src/ring/mod.rs [hunk 2]
```diff
@@ -422,6 +457,7 @@ impl Ring {
         // if the peer is just starting wait a bit before
         // we even attempt acquiring more connections
         tokio::time::sleep(Duration::from_secs(2)).await;
+        tracing::info!("Connection maintenance task: initial sleep completed");
```
Comment (iduartgomez) [inline crates/core/src/ring/mod.rs]: nit: debug

**Comment by sanity** [comment]
> @sanity Doesn't the message flow already roughly satisfy this though? The routing uses `k_closest_potentially_caching` to pick the next hop, so messages naturally move toward peers closer to the contract. The flow-based approach just "freezes" that routing decision at subscription time, capturing the path that actually worked (with skip lists, available connections, ...)

Yes, I think your approach is good

File: crates/core/src/ring/mod.rs [hunk 3]
```diff
@@ -450,11 +486,11 @@ impl Ring {
             let active_count = live_tx_tracker.active_connect_transaction_count();
             if let Some(ideal_location) = pending_conn_adds.pop_first() {
                 if active_count < MAX_CONCURRENT_CONNECTIONS {
-                    tracing::debug!(
+                    tracing::info!(
```
Comment (iduartgomez) [inline crates/core/src/ring/mod.rs]: nit: why the change in level

File: crates/core/src/ring/mod.rs [hunk 4]
```diff
@@ -450,11 +486,11 @@ impl Ring {
...
                         active_connections = active_count,
                         max_concurrent = MAX_CONCURRENT_CONNECTIONS,
-                        target_location = %ideal_location,
-                        "Attempting to acquire new connection"
+                        "Attempting to acquire new connection for location: {:?}",
```
Comment (iduartgomez) [inline crates/core/src/ring/mod.rs]: I actually prefer the structured message

File: crates/core/src/ring/mod.rs [hunk 5]
```diff
@@ -467,20 +503,19 @@ impl Ring {
...
                             );
                             error
                         })?;
                     if tx.is_none() {
                         let conns = self.connection_manager.connection_count();
                         tracing::warn!(
-                            connections = conns,
-                            target_location = %ideal_location,
-                            "acquire_new returned None - likely no peers to query through"
+                            "acquire_new returned None - likely no peers to query through (connections: {})",
```
Comment (iduartgomez) [inline crates/core/src/ring/mod.rs]: same, keep the logs structured where possible

File: crates/core/src/node/network_bridge/p2p_protoc.rs [hunk 6]
```diff
@@ -75,6 +75,40 @@ impl P2pBridge {
...
+        &self,
+        notifications: Vec<(freenet_stdlib::prelude::ContractKey, PeerKeyLocation)>,
+    ) {
+        if notifications.is_empty() {
+            return;
+        }
+
+        let own_location = self.op_manager.ring.connection_manager.own_location();
+
+        for (contract_key, upstream) in notifications {
```
Comment (iduartgomez) [inline crates/core/src/node/network_bridge/p2p_protoc.rs]: is possible to do this concurrently and not awaiting for each send sequentially?

File: crates/core/src/ring/mod.rs (hunk 3 above)
Comment (netsirius) [reply crates/core/src/ring/mod.rs]: I'll change it, I lost the change from main when I did the rebase

**Review by iduartgomez — APPROVED** [review]
Just fix conflicts and we can merge

**Comment by netsirius** [comment]
@claude review last changes

**Review by iduartgomez — APPROVED** [review]
